LibraBFT/Impl/Consensus/SafetyRules/Properties/SafetyRules.agda

```diff
@@ -43,18 +43,55 @@ module verifyAndUpdatePreferredRoundMSpec (quorumCert : QuorumCert) (safetyData
 
   safetyData' = safetyData & sdPreferredRound ∙~ twoChainRound
 
-  -- Before proving this, we should consider whether to add explicit support for <-cmp to our RWST
-  -- support, to make this proof unroll more "automatically".
+module verifyAndUpdatePreferredRoundMSpec (quorumCert : QuorumCert) (safetyData : SafetyData) where
+  open verifyAndUpdatePreferredRoundDefs quorumCert safetyData
 
-  postulate
-    contract
-      : ∀ P pre
-        → ((1cr<pr : C₁) → P (inj₁ fakeErr) pre [])
-        → ((1cr≥pr : ¬ C₁)
-           → ((2cr>pr : C₂) → P (inj₂ safetyData') pre [])
-              × ((2cr<pr : C₃) → P (inj₂ safetyData) pre [])
-              × ((2cr=pr : C₄) → P (inj₂ safetyData) pre []))
-        → LBFT-weakestPre (verifyAndUpdatePreferredRoundM quorumCert safetyData) P pre
+  module _ (pre : RoundManager) where
+    record setPR (sd : SafetyData) : Set where
+      field
+        eff   : sd ≡ safetyData'
```

Comment:
This won't be needed any time soon I think, but in principle we should be tracking the condition under which we update the preferred round (i.e., `twoChainRound > preferredRound`) -- otherwise how can we prove e.g. that the preferred round is nondecreasing? Maybe leave a TODO here for this and the corresponding condition (`twoChainRound <= preferredRound`) for `noChanges`.

Reply:
Thanks @cwjnkins, good catch, especially because addressing it revealed that I had proved it as if we update when `twoChainRound = preferredRound`, which isn't true.  FIXED.